logic_mod/Blocks/Api/CpuMath.cs

```diff
@@ -114,9 +153,246 @@ public object Atan(VarCtx ctx, object[] x)
+        //      convenient. If it is null (default), then the normal Newton-Raphson
+        //      or the secant method is used. If it is not null, then Halley's method
+        //      is used.
+        // x1 : float, optional
+        //      Another estimate of the zero that should be somewhere near the
+        //      actual zero. Used if `fprime` is not provided.
+        // rtol : float, optional
+        //      Tolerance (relative) for termination.
+        // full_output : bool, optional
+        //      If `full_output` is false (default), the root is returned.
+        //      If true, the dictionary {{"root": root}, {"converged": true/false},
+        //      {"iter": numIter}} is returned.
+        public object Newton(VarCtx ctx, object[] x)
+        {
+            int l = x.Length;
+
+            // Arguments and their default values:
+            object func;
+            float x0;
+            object fprime = null;
+            float tol = 1.48e-08F;
+            long maxiter = 50;
+            object fprime2 = null;
+            float x1 = 0;
+            float rtol = 0.0F;
+            bool full_output = false;
+
+            bool x1Provided = false;
+
+            if (l < 2 || l > 9)
+                throw new Exception("Invalid value");
+
+            // Conditionally initialize the arguments
+            int curArgIndex = 0;
+
+            func = x[curArgIndex++];
+
+            if (!BlockUtils.TryGetFloat(x[curArgIndex++], out x0))
+                throw new Exception("Invalid value");
+
+            if (curArgIndex >= l)
+                goto finishInitialization;
+            fprime = x[curArgIndex++];
+
+            if (curArgIndex >= l)
+                goto finishInitialization;
+            if (!BlockUtils.TryGetFloat(x[curArgIndex++], out tol))
+                throw new Exception("Invalid value");
+
+            if (curArgIndex >= l)
+                goto finishInitialization;
+            if (!BlockUtils.TryGetLong(x[curArgIndex++], out maxiter))
+                throw new Exception("Invalid value");
+
+            if (curArgIndex >= l)
+                goto finishInitialization;
+            fprime2 = x[curArgIndex++];
+
+
+            if (curArgIndex >= l)
+                goto finishInitialization;
+            x1Provided = BlockUtils.TryGetFloat(x[curArgIndex++], out x1);
+
+            if (curArgIndex >= l)
+                goto finishInitialization;
+            if (!BlockUtils.TryGetFloat(x[curArgIndex++], out rtol))
+                throw new Exception("Invalid value");
+
+            if (curArgIndex >= l)
+                goto finishInitialization;
+            if (!BlockUtils.TryGetBool(x[curArgIndex++], out full_output))
+                throw new Exception("Invalid value");
+
+            finishInitialization:
+            if (tol <= 0)
+                throw new Exception("tol too small (" + tol + " <= 0)");
+            if (maxiter < 1)
```

Comment:
Maybe max number of iterations should have some max value too? Otherwise user can pass 9999999999999999999999 here and CPU will never exit this function

Reply:
I think we should leave this to the user. If they decide to pass in a value that is very big then they should be allowed to. Also passing in a huge value would not break the game; since the numerical solver is declared as asynchronous, the game will continue and the only problem is that the CPU will be stuck here.